Add nfsv4-specific permissions checks to kernel

anodos325 opened the pull request:
There are various places in which evaluation of permissions
in the presence of an NFSv4 ACL is more nuanced than what is
typical when evaluating traditional POSIX permissions. For
example, a user may be permitted to delete a file if he
has DELETE permissions on the file or DELETE_CHILD permissions
on the parent directory. Traditional POSIX permissions will
only check for MAY_WRITE | MAY_EXEC on parent directory.

Several new inode permissions masks have been added to facilitate these
NFSv4-specific checks corresponding to different NFSv4 permissions
that grant abilities to make changes to files. For the purpose of
this commit and the goal of providing rough a approximation of
NFSv4 access checks, only write (and not read) access checks have
been implemented. This is selectively done in a way to grant
minimal compliance with permissions as defined in RFC-5661.

The new permissions-related behavior is only applied when the
inode sb_flag SB_NFS4ACL is present. In this case, the onus of full
implementation of requisite features to satisfy the ACL behavior
specified in RFC-5661 is delegated to the filesystem's inode
permissions interface (i_op->permission).

anodos325 committed:
Add nfsv4-specific permissions checks to kernel

There are various places in which evaluation of permissions
in the presence of an NFSv4 ACL is more nuanced than what is
typical when evaluating traditional POSIX permissions. For
example, a user may be permitted to delete a file if he
has DELETE permissions on the file or DELETE_CHILD permissions
on the parent directory. Traditional POSIX permissions will
only check for MAY_WRITE | MAY_EXEC on parent directory.

Several new inode permissions masks have been added to facilitate these
NFSv4-specific checks corresponding to different NFSv4 permissions
that grant abilities to make changes to files. For the purpose of
this commit and the goal of providing rough a approximation of
NFSv4 access checks, only write (and not read) access checks have
been implemented. This is selectively done in a way to grant
minimal compliance with permissions as defined in RFC-5661.

The new permissions-related behavior is only applied when the
inode sb_flag SB_NFS4ACL is present. In this case, the onus of full
implementation of requisite features to satisfy the ACL behavior
specified in RFC-5661 is delegated to the filesystem's inode
permissions interface (i_op->permission). If possible we try to
check for the convention POSIX permission first before trying
the NFSv4-equivalent. For example, when writing an xattr, we
check for WRITE_DATA before WRITE_NAMED_ATTRS because in the case
of former with a trivial ACL we can avoid having to evaluate the
full ACL, and instead merely look at POSIX mode.

anodos325 commented on fs/attr.c:
```diff
@@ -98,8 +144,27 @@ int setattr_prepare(struct dentry *dentry, struct iattr *attr)
+		 * Check for ACE4_WRITE_ATTRIBUTES. RFC 5661 Section 6.2.1.3.1
+		 * Permission to change times associated with file or directory
+		 * to an arbitrary value.
+		 */
```
I'll need to double-check this one. IIRC there is a nuance where WRITE_DATA / WRITE_ATTRIBUTES allows setting to current time, but not arbitrary time.

rick-mesta commented on fs/attr.c:
```diff
@@ -78,17 +78,63 @@ int setattr_prepare(struct dentry *dentry, struct iattr *attr)
+		if (!IS_NFSV4ACL(inode)) {
+			return -EPERM;
+		}
+		else if (inode_permission(inode, MAY_WRITE_OWNER)) {
```
I looked thru the [spec](https://tools.ietf.org/html/rfc5661#section-6.2.1.3.2), but couldn't find any differences w.r.t. `owner `vs `owner_group`. Is `MAY_WRITE_OWNER` the correct internal ACL equivalent flag even if you're manipulating the `gid` ?

anodos325 replied:
I'm actually adding MAY_WRITE_OWNER in this commit (didn't exist before). 

Mask gets passed to zpl_permissions() in ZFS, which then performs zfs_access() check for ACE_WRITE_OWNER. As you mentioned, ACE_WRITE_OWNER gives permissions to both chown() and chgrp().

anodos325 replied:
See here: https://github.com/truenas/zfs/pull/9/commits/85dcd0cb10ec1564e86f2f0f74d7a902ee26be06

rick-mesta commented on fs/attr.c:
```diff
@@ -244,7 +309,19 @@ int notify_change(struct dentry * dentry, struct iattr * attr, struct inode **de
 		if (!inode_owner_or_capable(inode)) {
+#if CONFIG_TRUENAS
+			if (IS_NFSV4ACL(inode)) {
+				error = inode_permission(inode, MAY_WRITE);
```
So, the assumption is, if you've got `MAY_WRITE`, you can not only modify the data but attributes as well. However, the inverse is not necessarily true, correct ?

anodos325 replied:
Correct. Issue here is actually the spec states that user having either ACE4_WRITE_DATA or ACE4_WRITE_ATTRIBUTES allows user to set time to _current_ server time. Doesn't allow arbitrary times. That's why it's permitted here, but not above. I should probably make this checks more explicit though (error == EACCES).

rick-mesta commented on fs/xattr.c:
```diff
@@ -250,8 +266,19 @@ __vfs_setxattr_locked(struct dentry *dentry, const char *name,
+#if CONFIG_TRUENAS
+	if (IS_NFSV4ACL(inode)) {
+		error = xattr_permission(inode, name, MAY_WRITE);
+		if (error) {
```
Same question as before in `fs/attr.c:314` 😏

anodos325 replied:
Same general answer. FreeBSD actually checks only WRITE_DATA. We check ACE4_WRITE_DATA then ACE4_WRITE_NAMED_ATTRS. 

There's a potential performance impact any time we start evaluating permissions that aren't ACE4_READ_DATA, ACE4_WRITE_DATA, and ACE4_EXECUTE. If mask can be represented by POSIX permissions, then we can check the znode pflags for whether the ACL is trivial. If it is trivial, then we know that the permissions can be fully expressed by the POSIX mode and bypass a zfs_access() call in favor of generic_permission().

rick-mesta commented on include/linux/fs.h:
```diff
@@ -102,6 +102,22 @@ typedef int (dio_iodone_t)(struct kiocb *iocb, loff_t offset,
+ * for permissions checks where NFSv4 ACL handling is
+ * more nuanced than a simple POSIX permissions.
+ */
+#define MAY_WRITE_NAMED_ATTRS	0x00000100
```
[nit]: please feel free to ignore any of my nits, but I find that having the hex values aligned at the next logical tab makes the code look cleaner.  Just my $0.02, fwiw

anodos325 replied:
I'll do that.

rick-mesta reviewed (approved):
ship it 🚢

gdevenyi commented:
Any reason this can't be upstreamed?

ghost commented:
~~We intend to.~~ Sorry, thought this was the zfs repo. It would be nice to get this upstreamed. Linux has a storied past of rejecting NFSv4 ACL support in various forms as well as indifference towards ZFS. It's rather discouraging, so nobody is pursuing upstreaming this at the moment.
FWIW, @anodos325 tells me the commonly used functionality is available through the ZFS changes alone, and these patches to the kernel are only needed to provide correct behavior of some NFSv4-specific features.

gdevenyi commented:
I would hope since these are NFSv4 ACL fixes, which touch the classic linux filesystem code, it should be acceptable without any mention of ZFS support :+1:

Q: how does this interact with the SMB ACLs? Do they already implement all this functionality on the kernel level?

anodos325 commented:
> Q: how does this interact with the SMB ACLs? Do they already implement all this functionality on the kernel level?

SMB ACLs are implemented through Samba's VFS, vfs_nfs4_acl_xattr. OS changes are NFSv4 ACL implementation agnostic. Changes come in two ways:
1) added inode SB_FLAG to indicate that NFSv4 ACLs are supported
2) added NFSv4-aware permissions (MAY_DELETE, etc) checks where appropriate / different from traditional Unix permissions (for example delete behavior). 

In principle, (1) and (2) could be implemented by other filesystems, which is not particularly onerous as (1) is already required for POSIX ACL support, and (2) would just require adding a i_op->permissions function.
